ntb_hw_switchtec.c

```diff
@@ -1573,6 +1617,146 @@ static int switchtec_ntb_reinit_peer(struct switchtec_ntb *sndev)
 	return rc;
 }
 
+struct ntb_sysfs_entry {
+	struct attribute attr;
+	ssize_t (*show)(struct switchtec_ntb *, char *);
+	ssize_t (*store)(struct switchtec_ntb *, const char *, size_t);
+};
+
+static ssize_t requester_ids_show(struct switchtec_ntb *sndev, char *page)
+{
+	int i;
+	int req_id_table_size;
+	char req_id_str[32];
+	u32 req_id;
+	ssize_t n = 0;
+
+	req_id_table_size = ioread16(&sndev->mmio_self_ctrl->req_id_table_size);
+
+	for (i = 0; i < req_id_table_size; i++) {
+		req_id = ioread32(&sndev->mmio_self_ctrl->req_id_table[i]);
+
+		if (!(req_id & NTB_CTRL_REQ_ID_EN))
+			break;
+
+		req_id >>= 16;
+		n += sprintf(req_id_str, "Requester ID #%d: %02x:%02x.%x\n", i,
+			     PCI_BUS_NUM(req_id), PCI_SLOT(req_id),
+			     PCI_FUNC(req_id));
+		strcat(page, req_id_str);
+	}
+
+	return n;
+}
+
+static ssize_t requester_ids_store(struct switchtec_ntb *sndev,
+				   const char *page, size_t count)
+{
+	int bus, dev, func;
+	int req_id;
+	bool add_req_id;
+	ssize_t ret;
+
+	if (sscanf(page, "add %x:%x.%x", &bus, &dev, &func) == 3) {
+		add_req_id = true;
+	} else if (sscanf(page, "del %x:%x.%x", &bus, &dev, &func) == 3) {
```

Comment:
I think this goes a bit against convention. There should be sysfs attributes for adding and removing. Similar to the bind and unbind attributes for attaching drivers. Better than having one attribute that does everything.

Reply:
I'll split it into three attributes 'add_requester_id', 'del_requester_id', and 'requester_ids' for adding, removing and showing.